doc/reference/kernel/timing/clocks.rst

```diff
@@ -1,161 +1,242 @@
+Apps with precise timing requirements (that are willing to do their
+own conversions to portable real time units) may access this with
+``k_uptime_ticks()``.
+
+Timeouts
+========
+
+The Zephyr kernel provides many APIs with a "timeout" parameter.
+Conceptually, this indicates the time at which an event will occur.
+For example:
+
+* Kernel blocking operations like ``k_sem_take()`` or
+  ``k_queue_get()`` may provide a timeout after which the routine will
+  return with an error code if no data is available.
+
+* Kernel ``struct k_timer`` objects must specify delays for their
+  duration and period.
+
+* The kernel ``k_delayed_work`` API provides a timeout parameter
+  indicating when a work queue item will be added to the system queue.
+
+All these values are specified using a ``k_timeout_t`` value.  This is
+an opaque struct type that must be initialized using one of a family
+of ``K_TIMEOUT`` macros.  The most common, ``K_TIMEOUT_MS()``, defines
+a time in milliseconds after the current time (strictly: the time at
+which the kernel receives the timeout value).
+
+Other options for timeout initialization follow the unit conventions
+described above: ``K_TIMEOUT_US()``, ``K_TIMEOUT_TICKS()`` and
+``K_TIMEOUT_CYC()`` specify timeout values that will expire after
+specified numbers of microseconds, ticks and cycles, respectively.
+
+Precision of ``k_timeout_t`` values is configurable, with the default
+being 32 bits.  Large uptime counts in non-tick units will experience
+complicated rollover semantics, so it is expected that
+timing-sensitive applications with long uptimes will be configured to
+use a 64 bit timeout type.
+
+Finally, it is possible to specify timeouts as absolute times since
+system boot.  A timeout initialized with ``K_TIMEOUT_UPTIME_MS()``
+indicates a timeout that will expire after the system uptime reaches
+the specified value.  There are likewise microsecond, cycles and ticks
+variants of this API.
+
+Timing Internals
+================
+
+Timeout Queue
+-------------
+
+All Zephyr ``k_timeout_t`` events specified using the API above are
+managed in a single, global queue of events.  Each event is stored in
+a double-linked list, with an attendant delta count in ticks from the
+previous event.  The action to take on an event is specified as a
+callback function pointer provided by the subsystem requesting the
+event, along with a ``struct _timeout`` tracking struct that is
+expected to be embedded within subsystem-defined data structures (for
+example: a ``struct wait_q``, or a ``k_tid_t`` thread struct).
+
+Note that all variant units passed via a ``k_timeout_t`` are converted
+to ticks once on insertion into the list.  There no
+multiple-conversion steps internal to the kernel, so precision is
+guaranteed at the tick level no matter how many events exist or how
+long a timeout might be.
+
+Note that the list structure means that the CPU work involved in
```

Comment:
Change this first sentence to:
```
Because of this timeout list structure, the CPU work to manage
large numbers of timeouts is quadratic to the number of active
timeouts.
```